lib/phoenix_live_view/async.ex

```diff
@@ -0,0 +1,214 @@
+    {:current_stacktrace, stack} = Process.info(self(), :current_stacktrace)
+    trapped_result = {:catch, :exit, reason, stack}
+    handle_async(socket, maybe_component, kind, keys, ref, trapped_result)
+  end
+
+  defp handle_kind(socket, maybe_component, :start, keys, result) do
+    callback_mod = maybe_component || socket.view
+
+    normalized_result =
+      case result do
+        {:ok, result} -> {:ok, result}
+        {:catch, kind, reason, stack} -> {:exit, to_exit(kind, reason, stack)}
+      end
+
+    case callback_mod.handle_async(keys, normalized_result, socket) do
+      {:noreply, %Socket{} = new_socket} ->
+        new_socket
+
+      other ->
+        raise ArgumentError, """
+        expected #{inspect(callback_mod)}.handle_async/3 to return {:noreply, socket}, got:
+
+            #{inspect(other)}
+        """
+    end
+  end
+
+  defp handle_kind(socket, _maybe_component, :assign, keys, result) do
+    case result do
+      {:ok, {:ok, %{} = assigns}} ->
+        new_assigns =
+          for {key, val} <- assigns do
+            {key, AsyncResult.ok(get_current_async!(socket, key), val)}
+          end
+
+        Phoenix.Component.assign(socket, new_assigns)
+
+      {:ok, {:error, reason}} ->
+        new_assigns =
+          for key <- keys do
+            {key, AsyncResult.error(get_current_async!(socket, key), reason)}
+          end
+
+        Phoenix.Component.assign(socket, new_assigns)
+
+      {:catch, kind, reason, stack} ->
+        normalized_exit = to_exit(kind, reason, stack)
+
+        new_assigns =
+          for key <- keys do
+            {key, AsyncResult.exit(get_current_async!(socket, key), normalized_exit)}
+          end
+
+        Phoenix.Component.assign(socket, new_assigns)
+    end
+  end
+
+  # handle race of async being canceled and then reassigned
+  defp prune_current_async(socket, keys, ref) do
+    case get_private_async(socket, keys) do
+      {^ref, _pid, _kind} -> {:ok, update_private_async(socket, &Map.delete(&1, keys))}
+      {_ref, _pid, _kind} -> :error
+      nil -> :error
+    end
+  end
+
+  defp update_private_async(socket, func) do
+    existing = socket.private[:phoenix_async] || %{}
+    Phoenix.LiveView.put_private(socket, :phoenix_async, func.(existing))
+  end
+
+  defp get_private_async(%Socket{} = socket, key) do
+    socket.private[:phoenix_async][key]
+  end
+
+  defp get_current_async!(socket, key) do
+    # handle case where assign is temporary and needs to be rebuilt
+    case socket.assigns do
+      %{^key => %AsyncResult{} = current_async} -> current_async
+      %{^key => _other} -> AsyncResult.new(key, key)
+      %{} -> raise ArgumentError, "missing async assign #{inspect(key)}"
+    end
+  end
+
+  defp to_exit(:throw, reason, stack), do: {{:nocatch, reason}, stack}
+  defp to_exit(:error, reason, stack), do: {reason, stack}
+  defp to_exit(:exit, reason, _stack), do: reason
+
+  defp cancel_existing(%Socket{} = socket, key) do
```

Comment:
might be worth making this public too, then someone can choose if they want to cancel an existing async and start a new one. a-la https://github.com/phoenixframework/phoenix_live_view/pull/2763/files#r1296905976